### ISSUE TITLE
Push Notifications Fail After App Reinstallation

### ISSUE THREAD

**Issue by JCsplash** [firebase/firebase-ios-sdk#14246]
### Description

Description
After deleting and reinstalling the app, Firebase Cloud Messaging (FCM) push notifications fail to work until the app is quit and restarted once or twice. This issue seems to be related to the APNs token not being set before retrieving the FCM token.

Error Messages
The following errors are logged during the issue:
`The operation couldn’t be completed. No APNS token specified before fetching FCM Token`
`APNS device token not set before retrieving FCM Token for Sender ID. Be sure to re-retrieve the FCM token once the APNS device token is set.`
`Declining request for FCM Token since no APNS Token specified`

**Steps to Reproduce**
1. Log in to the app and confirm that FCM notifications are working.
2. Delete the app.
3. Reinstall the app.
4. Observe that FCM notifications do not work immediately after reinstallation.

If you quit and restart the app multiple times, the notifications begin to work.

Code Snippet
This is the code used to register for remote notifications:

```
func registerForRemoteNotifications() {
    let application: UIApplication = UIApplication.shared
    if #available(iOS 10.0, *) {
        // For iOS 10 display notification (sent via APNS)
        UNUserNotificationCenter.current().delegate = (application.delegate) as? AppDelegate

        let authOptions: UNAuthorizationOptions = [.alert, .badge, .sound]
        UNUserNotificationCenter.current().requestAuthorization(
            options: authOptions,
            completionHandler: {_, _ in })
    } else {
        let settings: UIUserNotificationSettings =
            UIUserNotificationSettings(types: [.alert, .badge, .sound], categories: nil)
        application.registerUserNotificationSettings(settings)
    }
    application.registerForRemoteNotifications()
}
```

Environment
Firebase SDK Version: Upgraded from 8.8.0 to 11.6.0
Xcode 16.1
Swizzling: Enabled
Push Notification Settings: Configured as per documentation
Expected Behavior
After deleting and reinstalling the app, push notifications should work immediately without requiring multiple restarts of the app.

Actual Behavior
Push notifications fail to work immediately after reinstalling the app. The following errors are logged until the app is restarted once or twice:

`The operation couldn’t be completed. No APNS token specified before fetching FCM Token`
`APNS device token not set before retrieving FCM Token for Sender ID`
`Declining request for FCM Token since no APNS Token specified`

This issue did not occur with Firebase SDK version 8.8.0 but surfaced after upgrading to version 11.6.0.

### Reproducing the issue

**Steps to Reproduce**
1. Log in to the app and confirm that FCM notifications are working.
2. Delete the app.
3. Reinstall the app.
4. Observe that FCM notifications do not work immediately after reinstallation.

If you quit and restart the app multiple times, the notifications begin to work.

### Firebase SDK Version

11.6.0

### Xcode Version

16.1

### Installation Method

CocoaPods

### Firebase Product(s)

Messaging

### Targeted Platforms

iOS

### If using Swift Package Manager, the project's Package.resolved

<details>
<summary>Expand <code>Package.resolved</code> snippet</summary>
<br>

```json

Replace this line with the contents of your Package.resolved.

```

</details>

### If using CocoaPods, the project's Podfile.lock

<details>
<summary>Expand <code>Podfile.lock</code> snippet</summary>
<br>

```yml

Replace this line with the contents of your Podfile.lock!

```

</details>

**Comment by rizafran** [comment]
Hi @JCsplash, your issue seems related to https://github.com/firebase/firebase-ios-sdk/issues/10679#issuecomment-1402776795, especially that you've recently updated from a Firebase version before the change mentioned went into effect.

You may also consider checking [this](https://stackoverflow.com/a/69316428) troubleshooting guide.

**Comment by JCsplash** [comment]
Hey @rizafran why is v10.4.0+ still providing a non-nil `fcmToken` in the `didReceiveRegistrationToken` FCM delegate method? It seems like that value is refreshed only when an apnsToken is received but why is an invalid one being provided at app start time even before `UNUserNotificationCenter.current().requestAuthorization` and `UIApplication.shared.registerForRemoteNotifications()` are called? On that note, why before `apnsToken` is given, `Messaging.messaging().fcmToken` gives me an non-nil token that is invalid. Shouldn't this method return nil and the `didReceiveRegistrationToken` FCM delegate method not fire? Or can you explain why an invalid token is sent back?

**Comment by JCsplash** [comment]
Hi @leojaygoogle I also get the same error `No APNS token specified before fetching FCM Token.` which is expected - however when I call `Messaging.messaging().fcmToken` I am still getting a stale token. It's the old token before I uninstalled the app, while it should be nil.

**Comment by leojaygoogle** [comment]
Yeah, `fcmToken` returns a cached token until it's refreshed. https://github.com/firebase/firebase-ios-sdk/blob/75cc4df26d83f251629453ffae855c248bb6db3d/FirebaseMessaging/Sources/Token/FIRMessagingTokenManager.m#L78-L80

If you need the current token, please follow the doc: https://firebase.google.com/docs/cloud-messaging/ios/client#fetching-the-current-registration-token

**Comment by rizafran** [comment]
Hey @JCsplash, I'm wondering if you're still encountering the issue?
